rfcs/0011-separate-flyway-from-pipeline.md

```diff
@@ -0,0 +1,382 @@
+operations:
+
+BFD Pipeline Application:
+- Invokes Flyway on startup (which will execute any new migrations)
+- Runs hibernate validation on startup
+- Reads and writes data to the database
+
+BFD API Server Application:
+- Runs hibernate validation on startup
+- Reads data from the database (no writes)
+
+### Background on Database Migrations
+
+Within this proposal we will consider a database migration to be a group of one or more of the following types of
+changes that are deployed together:
+
+1. A database schema change consisting of one or more SQL migration scripts executed by Flyway
+2. An application change in the ORM layer (typically in relation to a change in the schema)
+
+When reasoning about different types of database migrations, it is necessary to consider that both the schema and the
+application may change. We will refer to the schema and application as they existed prior to the deployment as the old
+schema and the old application, and the schema and application as they exist after applying the migration as the new
+schema and the new applications. It is assumed that the old application is compatible with the old schema (since that is
+what is current running on the system) and that the new application is compatible with the new schema (ensuring this is
+a basic testing requirement of any change to the system). The interactions between old and new components though is
+important in understanding this proposal and leads to some classifications of migrations based on the compatibility of
+the schema change with old and new versions of the application:
+
+* A backward-compatible database migration is one where the new schema is compatible with the old applications
+(otherwise it is considered to be backward-incompatible)
+* A forward-compatible database migration is one where the old schema is compatible with the new applications
+(otherwise it is considered to be forward-incompatible)
+* A fully-compatible database migration is one that is both backward-compatible and forward-compatible
+* A fully-incompatible database migration is one that is both backward-incompatible and forward-incompatible
+
+Frequently, database migrations include both a schema change and an accompanying application change although there are
+some database migrations that consist only of a schema change. Strictly speaking, a database migration that does not
+contain a schema change is just an application change like any other application change. It is useful though for the
+purposes of this RFC for application changes deployed independently of a schema change to also be considered database
+migrations. Of interest in the next section is the fact that as defined above, a database migration that consists only
+of an application change is fully-compatible because it must work with both old and new schema (which are the same) to
+be a valid change at all.
```

Comment:
Moderate suggestion: A nuance here, and above, that I think would be helpful to clarify is that "old applications" and "new applications" are not... unbounded ranges. You really just need migrations that are forward- and backward-compatible with one deployed version. (Unless/until we move to a more complex deployment strategy that includes >2 application versions running simultaneously.)